File: include/Fc.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "Fc.h"
#include "helper.h"
 /**************************************************************************
 *PUBLIC FACING API
 **************************************************************************/
float* Fc(
            float* weight,
            int weight_w,
            float* bias,
            int bias_h, // bias_w = 1
            float* input,
            int relu_f
          )
{
  float* output = malloc(sizeof(float) * bias_h);

  for(int i = 0; i < bias_h; i++){
    output[i] = bias[i];
    for(int j = 0; j < weight_w; j++){
      output[i] += weight[i * weight_w + j] * input[j];
    }
    if(relu_f) {
      output[i] = relu(output[i]);
    }
  }
  free(input);
  free(weight);
  free(bias);
  return output;
}

File: golden/golden/fully_connected/fc.py
import sys
sys.path.append("../../")
from golden.basic_logic import addmult_v2 as alu
from golden.activation import relu

add = alu.bfloat_add
mult  = alu.bfloat_mult
format = alu.bfloat
relu = relu.functional

def fc(weight, weight_w, bias, bias_h, input, relu_f):
    output = [format('0') for i in range (bias_h)]
    for i in range(bias_h):
        output[i] = bias[i];
        for j in range(weight_w):
            output[i] = add(mult(weight[i * weight_w + j], input[j]), output[i])
        if(relu_f):
            output[i] = relu(output[i])
    return output
